### PR TITLE
TMOP + AMR + equal size optimization [tmop-amr-size]

### PR DESCRIPTION
**Pull request by vladotomov**
Taking into account local refinements when computing ideal_equal_size tmop targets.

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/895 | @vladotomov  | @tzanio | @rw-anderson + @kmittal2 | 11/23/19 | 12/4/19 | ⌛due 12/11/19 |

### CODE REVIEW

**Comment by kmittal2** [comment]
I did some tests and fixed a minor bug. We now get the same output if we optimize a regular mesh and its AMR counterpart (randomly-refined).

AMRvsRegular

**Comment by tzanio** [comment]
Ketan, can you click on "approve" when ready?

**Comment by kmittal2** [comment]
> Ketan, can you click on "approve" when ready?

Done.. I went through the PR last week and made some minor changes.

**Comment by rw-anderson** [comment]
> I did some tests and fixed a minor bug. We now get the same output if we optimize a regular mesh and its AMR counterpart (randomly-refined).
> 
> AMRvsRegular

Nice test.  Can this be repeated with some of the refinements being anisotropic?  Is that expected to work in this PR?

File: mesh/ncmesh.cpp [hunk 1]
```diff
@@ -4307,6 +4307,20 @@ int NCMesh::GetElementDepth(int i) const
    return depth;
 }
 
+int NCMesh::GetElementSizeReduction(int i) const
+{
+   int elem = leaf_elements[i];
+   int reduction = 1, parent;
```
Comment (rw-anderson) [inline mesh/ncmesh.cpp]: minor: maybe more clear that the comma functions as a separator and not an operator here if reversed:
int parent, reduction = 1;

Comment (kmittal2) [reply mesh/ncmesh.cpp]: Yes, I have addressed this now.

**Comment by tzanio** [comment]
@vladotomov, @kmittal2, @rw-anderson:

The regression testing caught a conflict between this PR and #799:

In this branch, the following works fine:
```
mesh-optimizer -m ./amr-quad-q2.mesh -o 2 -rs 1 -mid 9 -tid 2 -ni 200 -ls 2 -li 100 -bnd -qt 1 -qo 8 -no-vis
```

However, when this branch is combined with the one from #799, e.g. in `next`, the above produces the following error:
```
Verification failed: (x.Size() == Width()) is false:
 --> invalid input Vector size
 ... in function: const mfem::Vector &mfem::NonlinearForm::Prolongate(const mfem::Vector &) const
 ... in file: fem/nonlinearform.cpp:100
```

Can you please investigate and fix?

**Comment by kmittal2** [comment]
> Can you please investigate and fix?

I am almost done.. Once I have finalized my changes, should I push to [next](https://github.com/mfem/mfem/tree/next) or should I push to this branch?

**Comment by v-dobrev** [comment]
> > Can you please investigate and fix?
> 
> I am almost done.. Once I have finalized my changes, should I push to [next](https://github.com/mfem/mfem/tree/next) or should I push to this branch?

Push any changes in the branch first. Then, if the branch tests are good, it can be merged in `next` again.

**Comment by tzanio** [comment]
04e89ca424ac5c1b08ab3ff2c38af1b5085f237f fixed it, thanks!

**Comment by tzanio** [comment]
Great work, thanks @vladotomov and @kmittal2 !

**Pull request merged**